==========================================
Next-Gen Sequence Analysis Workshop (2016)
==========================================

.. Warning These documents are not maintained and their instructions may be
	out of date. However the GED Lab does maintain the `khmer protocols
	<http://khmer-protocols.readthedocs.org/>`__ which may cover similar
	topics. See also the `installation instructions for the current version
	of the khmer project <https://khmer.readthedocs.org/en/latest/install.html>`__.

This is the schedule for the `2016 MSU NGS course <http://bioinformatics.msu.edu/ngs-summer-course-2016>`__. All course communications will be organized around the `Slack channel <https://anguscourse.slack.com>`__.

.. which will run from August 8th to August 19st, 2016.  If you're interested in seeing the 2015 description, please see `the 2015 announcement <http://bioinformatics.msu.edu/ngs-summer-course-2015>`__.

This workshop has a :doc:`code-of-conduct`, do read it!

`Download all of these materials <https://github.com/ngs-docs/angus/archive/2016.zip>`__ or `visit the GitHub repository <https://github.com/ngs-docs/angus/tree/2016>`__.

Meal Times: Breakfast 7:45-8:45, Lunch 12-1, Dinner 6-7 (Unless noted below). See `HERE <http://conference.kbs.msu.edu/conference-center/food-and-beverage-service/>`__ for the menu and other info.

===============  =============================================================
Day              Schedule
===============  =============================================================
Monday 8/8       * 1:30pm lecture: `Welcome! <_static/2016-lecture1-welcome.pptx>`__ (Meg and Matt)
                 * 2:30pm: Getting Started with AWS :doc:`amazon/index` (Matt)
                 * 4pm: Intro to Linux :doc:`setup` and :doc:`MacManes_UNIX` and `Online Tutorial <http://rik.smith-unna.com/command_line_bootcamp/>`__ (Matt)
                 * 7pm tutorial, `background PDF BLAST <_static/BLAST.pdf>`__
                 * :doc:`running-command-line-blast` (Meg)
                 * 8pm: Instructor and TA Research presentations, socialize

Tuesday 8/9      * Tutorial :doc:`day2`
                 * 9:15am lecture: `Sequencing basics <_static/2016-lecture2-sequencing.pptx>`__ (Matt)
                 * 11:00am Assessment (Bob Drost)
                 * 1:15pm tutorial, `background PDF <_static/MacManes_Trimming.pdf>`__ and :doc:`MacManesTrimming` (Matt)
                 * 3:00pm Opionated guide to making a computational notebook. (Matt and other people)

                 * Evening *firepit social*

Wed 8/10         * 9:15am lecture, `mapping and variant calling lecture <_static/MappingAndVarCalling.pdf>`__ (Meg)
                 * 10:00am practical Variant Calling, :doc:`variant` (Meg)
                 * 1:15pm LinuxBrew `slides <http://sjackman.ca/linuxbrew-slides/#/>`__ (Shaun)
                 * 1:45pm practical, Installing Linuxbrew on AWS Ubuntu 14 :doc:`linuxbrew_install` (Torst)
                 * 2:15pm tutorial, BASH for genomics, :doc:`GenomicsShell` (Amanda)
                 * 7:00pm Mapping Quest :doc:`MacManes_MapQuest` (All)
                 * 8:30pm student presentations (Part 1 ~3 minutes each)

Thursday 8/11    * 9:15am lecture, `Intro to genome assembly <http://www.slideshare.net/torstenseemann/de-novo-genome-assembly-tseemann-imb-winter-school-2016-brisbane-au-4-july-2016>`__ (Torst)
                 * 10:15am `Genome assembly practical <http://sjackman.ca/abyss-activity/>`__  (Shaun)
                 * 1:15pm Genome assembly practical (continued)
                 * 2:15pm lecture, De Bruijn graph assembly (Shaun)
                 * 7:15pm Assembly challenge :doc:`MacManes_AssemblyQuest` (Matt)

Friday 8/12      * 9:15am `Debrief of Assembly Quest marathon <_static/AssemQuest_Debriefing.pptx>`__ (Matt)
                 * 9:30am lecture, `Bacterial genome annotation <_static/bacterial_genome_annotation.pdf>`__ (Torst)
                 * 10:00am practical Prokka :doc:`prokka_genome_annotation` (Torst)
                 * 11:00am practical Species identification with Kraken :doc:`kraken_species_identification` (Torst)
                 * 1:15pm tutorial, Intro to R :doc:`Rintro` (Amanda)
                 * 6:00pm Thai food take out at Mcrary Hall
                 * 7:30pm `Mars Rover Challenge! <http://sjackman.ca/mars-rover/>`__ (Shaun)
                 * 8:30pm Finish up student presentations

Saturday 8/13    * 9:15am Population genetics `lecture <_static/pop_genomics_singhal.pdf>`__ and practical (Sonal)
                 * 10:30am practical, :doc:`pop_gen_tutorial` (Sonal)
                 * 1:15pm lecture, `long read sequencing <_static/Torsten_Seemann_LRS.pdf>`__ (Torst)
                 * 2:00pm practical, :doc:`analyzing_nanopore_data` (Lisa)
                 * 6:00pm BBQ on the island if nice out, in the classroom if not

Sunday 8/14      * Free Day
                 * Brunch at McCrary 12pm - 1pm*
                 * Dinner field trip to Kalamazoo

Monday 8/15      * 9:15am , Transcriptome assembly & evaluation `lecture <_static/MacManes_transcriptome.pptx>`__ and  :doc:`MacManes_Trinity` (Matt)
                 * 1:15pm Alignment / Mapping `lecture <http://robpatro.com/redesign/AlignmentAndMapping.pdf>`__ and `practical <http://angus.readthedocs.io/en/2016/rob_map/tut.html>`__ (Rob)
                 * 6:30pm LATER DINNER TIME
                 * 7:45pm: Twitter, Blogging and Bioinformatics (All)
                 * 8:30pm: Finish up student presentations! (All)

Tuesday 8/16     * 9:15am Transcript quantification `lecture <http://robpatro.com/redesign/Quantification.pdf>`__ (Rob)
                 * 10:30am practical `Using Salmon for read counts <http://angus.readthedocs.io/en/2016/rob_quant/tut.html>`__ (Rob)
                 * 11:45pm lecture, `mRNAseq differential expression <_static/DifferentialExpressionBasics_NGS2016_ID.pdf>`__ (Ian)
                 * 12:30pm Lunch (new later time)
                 * 1:15pm `Visualizing RNA-Seq counts with Degust <http://www.vicbioinformatics.com/degust/>`__ (Torst)
                 * 1:45pm `mRNA stats <_static/NGS2015_RNAseq_ID_1.pptx>`__ (Ian)
                 * 2:45pm practical `Using DEseq2 for DGE <https://github.com/ngs-docs/angus/blob/2016/DeSeq2_Salmon_Tutorial.md>`__ (Ian)
                 * 6:00pm Dinner (back to original time)
                 * 7:15pm Optional :doc:`arabidopsis_assembly_challenge`

Wed 8/17         * 9:15am `lecture <_static/GetData.pdf>`__ and `Sourcing NCBI data including SRA practical <https://github.com/mestato/epp622/blob/master/ncbi.md>`__ (Meg)
                 * 10:15am `More fun with R <https://github.com/MingChen0919/KBS_workshop_Michigan_2016/blob/master/fun_with_R.md>`__ (Ming)
                 * 12:00 Lunch
                 * Bonus `Routes Through Differential Expression Analysis <_static/RoutesThroughDESeqAnalysis.pdf>`__
                 * 1:00pm UNIX Stuff :doc:`MacManes_UNIX` (All) [tentative]
                 * 6:00pm Dinner
                 * 7:15pm Work on your own data, ask questions..

Thursday 8/18    * 9:15am Open source your scripts with GitHub (Lisa) :doc:`LC-github`
                 * 11:30 Jupyter Notebook (Lisa) :doc:`jupyter_notebook`
                 * 1:15pm Assessment (Bob Drost)
                 * Meet outside classroom after assessment for group photo
                 * 2pm: `Functional Annotation slides <_static/FunctionalAnnotation.pdf>`__ and :doc:`functional_annotation_2016` (Meg)
                 * 6:00pm Dinner
                 * Evening: Ask the Expert (All)

Friday 8/19      * 9:15-9:45 `Closing Lecture! <_static/2016-final-lecture.pptx>`__ (Meg and Matt)
                 * 10am discussion about class; more stuff


                 * Links:
                   `Opinionated guides to NGS <http://davis-assembly-masterclass-2013.readthedocs.org/en/latest/outputs/index.html>`__ /
                   `Software Carpentry <http://software-carpentry.org>`__
                   `Data Carpentry <http://datacarpentry.org>`__

===============  =============================================================

Dramatis personae
=================

Instructors:

* Meg Staton `@HardwoodGenomic <https://twitter.com/HardwoodGenomic?lang=en>`__.
* Matt MacManes `@macmanes <https://twitter.com/macmanes?lang=en>`__.
* Torsten Seemann `@torstenseemann <https://twitter.com/torstenseemann?lang=en>`__.
* Shaun Jackman `@sjackman <https://twitter.com/sjackman?lang=en>`__.
* Rob Patro `@nomad421 <https://twitter.com/nomad421?lang=en>`__.
* Ian Dworkin `@IanDworkin <https://twitter.com/IanDworkin?lang=en>`__.
* Sonal Singhal
* Amanda Charbonneau `@procrastinomics <https://twitter.com/procrastinomics?lang=en>`__.


TAs:

* Lisa Cohen `@monsterbashseq <https://twitter.com/monsterbashseq?lang=en>`__.
* Will Pitchers `@steeljawpanda <https://twitter.com/steeljawpanda?lang=en>`__.
* Ming Chen


She Who Drives Many Places:

* Kate MacManes

Genome Assembly References
==========================
* `Ben Langmead Resources <http://www.langmead-lab.org/teaching-materials/>`__
* `How do we assemble genomes (videos) <https://www.youtube.com/watch?list=PLQ-85lQlPqFNGdaeGpV8dPEeSm3AChb6L&v=vjB6nhOu3BY>`__


Papers and References
=====================

Books
-----

* `Bioinformatics Data Skills <http://shop.oreilly.com/product/0636920030157.do>`__
* `Practical Computing for Biologists <http://practicalcomputing.org/>`__

  These books, especially the 1st, are highly recommended book for people looking for a systematic
  presentation on shell scripting, programming, UNIX, etc.

RNAseq
------

* `Differential gene and transcript expression analysis of RNA-seq
  experiments with TopHat and Cufflinks
  <http://www.ncbi.nlm.nih.gov/pubmed/22383036>`__, Trapnell et al.,
  Nat. Protocols.

  One paper that outlines a pipeline with the tophat, cufflinks, cuffdiffs and
  some associated R scripts.

* `Statistical design and analysis of RNA sequencing
  data. <http://www.ncbi.nlm.nih.gov/pubmed/20439781>`__, Auer and
  Doerge, Genetics, 2010.

* `A comprehensive comparison of RNA-Seq-based transcriptome analysis from reads to differential gene expression and cross-comparison with microarrays: a case study in Saccharomyces cerevisiae. <http://www.ncbi.nlm.nih.gov/pubmed/?term=22965124>`__ Nookaew et al., Nucleic Acids Res. 2012.

* `Challenges and strategies in transcriptome assembly and differential gene expression quantification. A comprehensive in silico assessment of RNA-seq experiments <http://www.ncbi.nlm.nih.gov/pubmed/?term=22998089>`__ Vijay et al., 2012.

* `Computational methods for transcriptome annotation and quantification using RNA-seq <http://www.ncbi.nlm.nih.gov/pubmed/21623353>`__, Garber et al., Nat. Methods, 2011.

* `Evaluation of statistical methods for normalization and differential expression in mRNA-Seq experiments. <http://www.ncbi.nlm.nih.gov/pubmed/?term=20167110>`__, Bullard et al., 2010.

* `A comparison of methods for differential expression analysis of RNA-seq data <http://www.biomedcentral.com/1471-2105/14/91>`__, Soneson and Delorenzi, BMC Bioinformatics, 2013.

* `Measurement of mRNA abundance using RNA-seq data: RPKM measure is inconsistent among samples. <http://www.ncbi.nlm.nih.gov/pubmed/?term=22872506>`__, Wagner et al., Theory Biosci, 2012.  Also see `this blog post <http://blog.nextgenetics.net/?e=51>`__ explaining the paper in detail.

Computing and Data
------------------

* `A Quick Guide to Organizing Computational Biology Projects <http://www.ploscompbiol.org/article/info%3Adoi%2F10.1371%2Fjournal.pcbi.1000424>`__, Noble, PLoS Comp Biology, 2009.

* `Willingness to Share Research Data Is Related to the Strength of the Evidence and the Quality of Reporting of Statistical Results <http://software-carpentry.org/blog/2012/05/the-most-important-scientific-result-published-in-the-last-year.html>`__, Wicherts et al., PLoS One, 2011.

* `Got replicability? <http://econjwatch.org/articles/got-replicability-the-journal-of-money-credit-and-banking-archive>`__, McCullough, Economics in Practice, 2007.

Also see this great pair of blog posts on `organizing projects <http://nicercode.github.io/blog/2013-04-05-projects/>`__ and `research workflow <http://carlboettiger.info/2012/05/06/research-workflow.html>`__.

Links
=====

Humor
-----

* `Data Sharing and Management Snafu in 3 Short Acts <http://www.youtube.com/watch?v=N2zK3sAtr-4&feature=youtu.be>`__

Resources
---------

* `Biostar <http://biostars.org>`__

  A high quality question & answer Web site.

* `SEQanswers <http://seqanswers.com/>`__

  A discussion and information site for next-generation sequencing.

* `Software Carpentry lessons <http://software-carpentry.org/4_0/index.html>`__

  A large number of open and reusable tutorials on the shell, programming,
  version control, etc.

Blogs
-----

* http://www.genomesunzipped.org/

  Genomes Unzipped.

* http://ivory.idyll.org/blog/

  Titus's blog.

* http://bcbio.wordpress.com/

  Blue Collar Bioinformatics

* http://massgenomics.org/

  Mass Genomics

* http://blog.nextgenetics.net/

  Next Genetics

* http://gettinggeneticsdone.blogspot.com/

  Getting Genetics Done

* http://omicsomics.blogspot.com/

  Omics! Omics!

* http://lab.loman.net/

  Nick Loman's lab notebook

* http://TheGenomeFactory.blogspot.com/

  The Genome Factory (Torsten Seemann)

Material from previous years
============================

.. toctree::
   :maxdepth: 5

   day1
   day2
   variant

   assembling-ecoli-with-velvet
   interval-analysis-and-visualization
   sam-format-tutorial
   R_Introductory_tutorial_2014
   IntroductionControlFlowR
   vcf_exploration
   eel-pond

   SOAPdeNovoTrans_count_eXpress
   analyzing_drosophila_htseq
   drosophila_rnaseq1
   drosophila_rnaseq_bwa_htseq
   genome-comparison-and-phylogeny
   git-koans
   howe-mgrast
   howe-ncbi
   kmers-etc
   long-read
   mount_chris_snapshot
   short-read-quality-evaluation-ctb

   amazon/index
   instructors-guide
   code-of-conduct
